File: server/src/models/finance.rs
use mongodb::bson::{oid::ObjectId, DateTime};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Fee {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub fee_name: String,
    pub description: Option<String>,
    pub amount: f64,
    pub currency: String,            // "USD" | "KHR"
    pub grade_level: Option<String>, // None = all grades
    pub academic_year: String,
    pub due_date: DateTime,
    pub is_mandatory: bool,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

impl Fee {
    pub fn new(
        school_id: String,
        fee_name: String,
        amount: f64,
        currency: String,
        academic_year: String,
        due_date: DateTime,
    ) -> Self {
        let now = DateTime::now();
        Self {
            id: None,
            school_id,
            fee_name,
            description: None,
            amount,
            currency,
            grade_level: None,
            academic_year,
            due_date,
            is_mandatory: true,
            created_at: now,
            updated_at: now,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Payment {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub student_id: ObjectId,
    pub fee_id: ObjectId,
    pub amount_paid: f64,
    pub currency: String,
    pub payment_method: String, // "cash" | "bank_transfer" | "mobile_payment" | "card"
    pub payment_date: DateTime,
    pub receipt_number: String,
    pub status: String, // "pending" | "completed" | "failed"
    pub remarks: Option<String>,
    pub processed_by: ObjectId,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

impl Payment {
    pub fn new(
        school_id: String,
        student_id: ObjectId,
        fee_id: ObjectId,
        amount_paid: f64,
        currency: String,
        payment_method: String,
        receipt_number: String,
        processed_by: ObjectId,
    ) -> Self {
        let now = DateTime::now();
        Self {
            id: None,
            school_id,
            student_id,
            fee_id,
            amount_paid,
            currency,
            payment_method,
            payment_date: now,
            receipt_number,
            status: "completed".to_string(),
            remarks: None,
            processed_by,
            created_at: now,
            updated_at: now,
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Invoice {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub invoice_number: String,
    pub student_id: ObjectId,
    pub fee_ids: Vec<ObjectId>,
    pub total_amount: f64,
    pub amount_paid: f64,
    pub balance: f64,
    pub currency: String,
    pub issue_date: DateTime,
    pub due_date: DateTime,
    pub status: String, // "unpaid" | "partial" | "paid" | "overdue"
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

impl Invoice {
    pub fn new(
        school_id: String,
        invoice_number: String,
        student_id: ObjectId,
        fee_ids: Vec<ObjectId>,
        total_amount: f64,
        currency: String,
        due_date: DateTime,
    ) -> Self {
        let now = DateTime::now();
        Self {
            id: None,
            school_id,
            invoice_number,
            student_id,
            fee_ids,
            total_amount,
            amount_paid: 0.0,
            balance: total_amount,
            currency,
            issue_date: now,
            due_date,
            status: "unpaid".to_string(),
            created_at: now,
            updated_at: now,
        }
    }
}

File: dashboard/types/finance.ts
// Finance TypeScript types

export interface Fee {
	id: string;
	schoolId: string;
	feeName: string;
	description?: string;
	amount: number;
	currency: string;
	gradeLevel?: string;
	academicYear: string;
	dueDate: string;
	isMandatory: boolean;
	formattedAmount: string;
	isAllGrades: boolean;
	createdAt: string;
	updatedAt: string;
}

export interface Payment {
	id: string;
	schoolId: string;
	studentId: string;
	feeId: string;
	amountPaid: number;
	currency: string;
	paymentMethod: string;
	paymentMethodLabel: string;
	paymentDate: string;
	receiptNumber: string;
	status: string;
	remarks?: string;
	processedBy: string;
	formattedAmount: string;
	createdAt: string;
	updatedAt: string;
}

export interface Invoice {
	id: string;
	schoolId: string;
	invoiceNumber: string;
	studentId: string;
	feeIds: string[];
	totalAmount: number;
	amountPaid: number;
	balance: number;
	currency: string;
	issueDate: string;
	dueDate: string;
	status: string;
	formattedTotal: string;
	formattedBalance: string;
	paymentPercentage: number;
	isOverdue: boolean;
	createdAt: string;
	updatedAt: string;
}

export interface FinanceSummary {
	totalCollected: number;
	totalPending: number;
	totalOverdue: number;
	currency: string;
	studentsWithBalance: number;
	totalStudents: number;
	collectionRate: number;
}

// Input types
export interface CreateFeeInput {
	schoolId: string;
	feeName: string;
	description?: string;
	amount: number;
	currency: string;
	gradeLevel?: string;
	academicYear: string;
	dueDate: string;
	isMandatory?: boolean;
}

export interface UpdateFeeInput {
	feeName?: string;
	description?: string;
	amount?: number;
	currency?: string;
	gradeLevel?: string;
	academicYear?: string;
	dueDate?: string;
	isMandatory?: boolean;
}

export interface CreatePaymentInput {
	schoolId: string;
	studentId: string;
	feeId: string;
	amountPaid: number;
	currency: string;
	paymentMethod: string;
	remarks?: string;
}

export interface CreateInvoiceInput {
	schoolId: string;
	studentId: string;
	feeIds: string[];
	totalAmount: number;
	currency: string;
	dueDate: string;
}

export interface PaymentFilters {
	studentId?: string;
	status?: string;
	paymentMethod?: string;
	dateFrom?: string;
	dateTo?: string;
}

export interface InvoiceFilters {
	studentId?: string;
	status?: string;
}

// Fee form data for UI
export interface FeeFormData {
	feeName: string;
	description: string;
	amount: string;
	currency: "USD" | "KHR";
	gradeLevel: string;
	academicYear: string;
	dueDate: string;
	isMandatory: boolean;
}

// Payment form data for UI
export interface PaymentFormData {
	studentId: string;
	feeId: string;
	amountPaid: string;
	currency: "USD" | "KHR";
	paymentMethod: "cash" | "bank_transfer" | "mobile_payment" | "card";
	remarks: string;
}

// Payment method options
export const PAYMENT_METHODS = [
	{ value: "cash", label: "Cash" },
	{ value: "bank_transfer", label: "Bank Transfer" },
	{ value: "mobile_payment", label: "Mobile Payment (Wing/KHQR)" },
	{ value: "card", label: "Card (Visa/Mastercard)" },
] as const;

// Currency options
export const CURRENCIES = [
	{ value: "USD", label: "US Dollar (USD)", symbol: "$" },
	{ value: "KHR", label: "Khmer Riel (KHR)", symbol: "៛" },
] as const;

// Status options
export const PAYMENT_STATUSES = [
	{ value: "completed", label: "Completed", color: "success" },
	{ value: "pending", label: "Pending", color: "warning" },
	{ value: "failed", label: "Failed", color: "danger" },
] as const;

export const INVOICE_STATUSES = [
	{ value: "unpaid", label: "Unpaid", color: "default" },
	{ value: "partial", label: "Partial", color: "warning" },
	{ value: "paid", label: "Paid", color: "success" },
	{ value: "overdue", label: "Overdue", color: "danger" },
] as const;
